tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,235 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2019-2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "execute.hpp"
+#include "parser.hpp"
+#include <uvwasi.h>
+#include <cassert>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+
+namespace
+{
+// Global state.
+// NOTE: we do not care about uvwasi_destroy(), because it only clears file descriptors (currently)
+// and we are a single-run tool. This may change in the future and should reevaluate.
+uvwasi_t state;
+
+fizzy::execution_result wasi_return_enosys(fizzy::Instance&, std::vector<uint64_t>, int)
+{
+    return {false, {UVWASI_ENOSYS}};
+}
+
+fizzy::execution_result wasi_proc_exit(fizzy::Instance&, std::vector<uint64_t> args, int)
+{
+    uvwasi_proc_exit(&state, static_cast<uvwasi_exitcode_t>(args[0]));
+    // Should not reach this.
+    return {false, {}};
+}
+
+fizzy::execution_result wasi_fd_write(fizzy::Instance& instance, std::vector<uint64_t> args, int)
+{
+    uint32_t fd = static_cast<uint32_t>(args[0]);
+    uint32_t iovptr = static_cast<uint32_t>(args[1]);
+    uint32_t iovcnt = static_cast<uint32_t>(args[2]);
+    uint32_t nwritten = static_cast<uint32_t>(args[3]);
```

Comment:
Perhaps this is `nwrittenptr` and `nwritten_` is just `nwritten`
```suggestion
    uint32_t nwrittenptr = static_cast<uint32_t>(args[3]);
```

Reply:
I copied these from uvwasi, but perhaps we should consider following the naming of `wasi.witx`. That does not use `ptr` anywhere, we could decide to add `_ptr` to pointers.

What do you think?

Reply:
I think adding `_ptr` would be good.

Reply:
Fixed.